src/ScopedValues.jl

```diff
@@ -60,6 +60,22 @@ Test if the ScopedValue has a default value.
 """
 Base.isassigned(val::ScopedValue) = val.has_default
 
+struct Snapshot
+    values::Vector{Pair{ScopedValue, Any}}
+end
+Snapshot(values::ScopedValue...) = Snapshot(collect(map(v->v => v[], values)))
+
+"""
+    @snapshot exprs...
+
+Create a snapshot containing the `ScopedValue`(s) in `exprs`. This snapshot can
+be used in [`with`](@ref) or [`@with`](@ref) calls.
+"""
+macro snapshot(exprs...)
+    exprs = map(esc, exprs)
+    :($(Snapshot)($(exprs...)))
+end
```

Comment:
Why a macro, rather than a function?
```julia
sshot = snapshot(sval, sval_float)
```
I _think_ it could be a function?

Reply:
Yeah agreed, it makes probably more sense as a function. 

The macro form of `@with` exists purely to avoid the extra call-frame.